gui/console.cpp

```diff
@@ -620,13 +620,27 @@ void ConsoleDialog::saveHistory() {
 		return;
 	}
 
-	for (int i = 0; i < _historySize; ++i) {
-		saveFile->writeString(_history[i]);
-		saveFile->writeByte('\n');
+	// Saving the history entries in the proper order;
+	// The most recent entry should be the last to be saved.
+	// NOTE When the _history table is full, we need to start saving
+	//      from one slot after (in a circular manner) the _historyIndex slot.
+	//      In this case the _historyIndex slot contains the temporary stored user input,
+	//      which we do not want to persist.
+	//      This means that when full, (kHistorySize - 1) entries will be saved.
+	//      When the table is not full, storing always begins from index 0.
+	int idx = (kHistorySize == _historySize) ? ((_historyIndex + 1) % kHistorySize) : 0;
+	int entriesWritten = 0;
+	while (idx != _historyIndex) {
+		if (!_history[idx].empty()) {
```

Comment:
Instead of adding the loop logic here, too - we should maybe think about using a RingBuffer structure - I've written one here which could be used: https://github.com/mgerhardy/vengi/blob/master/src/modules/core/collection/RingBuffer.h

Then just use the iterators here.

Reply:
> Do we store empty history entries? I would rather fix that instead of checking it here, no?
This was mainly bug before when we could store the _historyIndex entry to the persistent file, since that entry was not properly added to history, just saved to the persistent file (by mistake) when the history buffer was full. So if it was empty, the persistent file would get an empty entry, and then upon load (since we check then for an empty string and break loading) it would possibly cut the loaded history short (not load the whole file as it was supposed to but stop at wherever the empty string was since we also did not store by order). 
We also could not check for the user input entry being empty before putting in the _historyIndex slot, because then it would potentially bring up a deleted user input which would confuse the user. I tried that before and it was pointed out to me that it was indeed not correct
<https://discord.com/channels/581224060529148060/711242520415174666/963728979062308884>

The bug is probably now fixed due to either of the rest of the changes in the PR (not include _historyIndex entry to the persistent file when the buffer is full, and save by proper order), but it doesn't hurt to keep the check there, I think.

I agree that we could use a ringbuffer for the history buffer, which would abstract some the logic of handling the "ring" with loops inside the console.cpp. However, like you said I'd prefer it if perhaps this could be the task of a future PR, rather than do it in this one.

Reply:
I agree with your comment. Checking if the history item is empty is probably not needed now that you no longer save the _historyIndex item, but it cannot hurt.

As for using a ring buffer, that can be done as an improvement in the future but is not needed for this pull request.